### PR TITLE
automatic detection of multioutput datasets

### PR DESCRIPTION
**Pull request by jhmenke**
## What does this PR do?

Slight variation of #903 depending on the shape of the target array.

## What are the relevant issues?

#971

### CODE REVIEW

**Comment by weixuanfu** [comment]
Neat! Thank you for this PR. I think it should work for checking multi output dataset. But I checked this [link](https://stats.stackexchange.com/questions/153853/regression-with-scikit-learn-with-multiple-outputs-svr-or-gbm-possible), I think that not all regressors in our default config are supporting multi-output regression. I think we need a workaround for this issue.

**Comment by jhmenke** [comment]
There are several ways to go about this:
- Give a warning if the standard regression / classification config is chosen for multi outputs, which would require the user to create their own config for multi output datasets.
- Make a list with regressors and classifiers that don't support multi output and remove them automatically if such a dataset is detected
- Embed these regressors and classifiers in MultiOutputRegressor/Classifier in the default config. Not sure about the performance losses there

**Comment by weixuanfu** [comment]
Thank you for your ideas here. I prefer the 3rd one but I think a better solution based on it is to automatically use `MultiOutputRegressor/Classifier` over pipelines that generated from current default config if TPOT detected the y is a multi-output target. Please let me know if you have any idea.

**Comment by jhmenke** [comment]
While working on it, i think i spotted a bug:

https://github.com/EpistasisLab/tpot/blob/aea42a59b52ff9704f9f677a01ebb71fda4520cc/tpot/operator_utils.py#L215


This should be *dep_import_str* as the key and not *import_str*, right? It seems this also changes some tests.

**Comment by jhmenke** [comment]
If i run the failing tests manually, they work. It has to be some unsuitable combination due to the random seed.. i'm not sure how to debug this.

**Comment by weixuanfu** [comment]
> 
> 
> While working on it, i think i spotted a bug:
> 
> https://github.com/EpistasisLab/tpot/blob/aea42a59b52ff9704f9f677a01ebb71fda4520cc/tpot/operator_utils.py#L215
> 
> This should be _dep_import_str_ as the key and not _import_str_, right? It seems this also changes some tests.

Yes, that is a bug. Hmm, I thought I fixed it a while ago but might not merge to master/dev branch.

File: tpot/base.py [hunk 1]
```diff
@@ -501,6 +500,35 @@ def _fit_init(self):
             self._last_optimized_pareto_front_n_gens = 0
             self._setup_config(self.config_dict)
 
+            if multi_output_target:
```
Comment (weixuanfu) [inline tpot/base.py:509]: Modifying _config_dict may not work in the situation that use use a customized configurations  instead of default one. So, I think a practical way is to modify the `_compile_to_sklearn` function ([here](https://github.com/EpistasisLab/tpot/blob/v0.11.1/tpot/base.py#L1193-L1194)). If `multi_output_target` is `True`, then
`sklearn_pipeline=MultiOutputClassifier(estimator=sklearn_pipeline)` or  `sklearn_pipeline=MultiOutputRegessor(estimator=sklearn_pipeline)` . I think it maybe a more general solution for multioutput dataset.

Comment (jhmenke) [reply tpot/base.py:509]: This seems better, to be honest i didn't find a good place where to put my code and only settled on the _fit_init function therefore.

Comment (jhmenke) [reply tpot/base.py:509]: Okay i looked into it, but the code would be a mess. Several functions would have to take multi_output_target as a new argument (most of them in export_utils.py), since they don't have access to the data or the TPOT Object. 

Imo _fit_init seems to be the least intrusive point to include the checks

Comment (weixuanfu) [reply tpot/base.py:509]: Thank you for looking into. You are right. I think TPOT exported codes should also include MultiOutputRegessor/MultiOutputClassifier, which should change a lot of codes in TPOT. I will look into it when I get some time next week.

Comment (jhmenke) [reply tpot/base.py:509]: any updates?

Comment (weixuanfu) [reply tpot/base.py:509]: @jhmenke I am sorry for overlooking this. I did not get a chance to look into this issue those days due to my busy schedule. I agree that TPOT need some major changes for including `MultiOutputRegessor/MultiOutputClassifier`. I may get some time in March to add those changes. You are welcome to push any changes meanwhile.

Comment (jhmenke) [reply tpot/base.py:509]: Could we use my PR as a temporary fix until you have time to thoroughly refactor the code? I can prepare an update with the current development branch.

Comment (weixuanfu) [reply tpot/base.py:509]: Sorry for the delay. I think we can use it for a temporary solution with a minor release.

Comment (jhmenke) [reply tpot/base.py:509]: Okay i merged the current master/development into this. Should be good to go as an interim solution then.

File: tpot/base.py [hunk 2]
```diff
@@ -507,6 +506,35 @@ def _fit_init(self):
...
+                    'sklearn.linear_model.ElasticNetCV',
+                    'sklearn.svm.LinearSVR',
+                    'xgboost.XGBRegressor',
+                    'sklearn.linear_model.SGDRegressor'
+                ]
+                for model in list(self._config_dict.keys()):
+                    if model in single_output_classifiers:
+                        if 'sklearn.multioutput.MultiOutputClassifier' not in self._config_dict.keys():
+                            self._config_dict['sklearn.multioutput.MultiOutputClassifier'] = {"estimator": {}}
+                        self._config_dict['sklearn.multioutput.MultiOutputClassifier']['estimator'][model] = self._config_dict[model]
```
Comment (weixuanfu) [inline tpot/base.py:527]: There is only one sklearn.multioutput.MultiOutputClassifier in self._config_dict and `estimator` is keeping updated until the last model in  `single_output_classifiers` so that the rest models should be removed.

**Review by weixuanfu — CHANGES_REQUESTED** [review]
I think we cannot add this support for now since it need more changes.

**Comment by jessegmeyerlab** [comment]
Is this the closest pull to multi-output? any plans to finish it?

**Comment by jhmenke** [comment]
> 
> 
> Is this the closest pull to multi-output? any plans to finish it?

sorry, no time for this right now..  feel free to work with my branch. i think the last todo is cloning the multioutputclassifier for every estimator that only supports single outputs as per weixuanfu's review.

**Comment by windowshopr** [comment]
Would love to see multioutput regression/classification working with tpot! Keep going! Haha
